### ISSUE TITLE
ui:repeat child state saving depends on the source of the request

### ISSUE THREAD

**Issue by javaserverfaces** [javaee/mojarra#4247]
State saving of components nested in ui:repeat behaves differently when it is caused by a partial Ajax triggered _inside_ the ui:repeat vs. a request outside the ui:repeat or a non-Ajax request (form submit).

Consider this basic ui:repeat loop:

```
<ui:repeat value="#{testBean.items}" var="row">
    <test:stateSavingComponent id="inRepeat" updateId=":form"/>
</ui:repeat>
```

test:stateSavingComponent is a composite component that contains two command links to update the component state (a counter):

```
<h:panelGrid columns="3">
    <h:outputText id="value" value="#{cc.clientId}: #{cc.counter}"/>

    <h:commandLink action="#{cc.increase}" value="+1 (ajax)">
        <f:ajax execute="#{cc.attrs.updateId}" render="#{cc.attrs.updateId}"/>
    </h:commandLink>

    <h:commandLink action="#{cc.increase}" value="+1 (submit)"/>
</h:panelGrid>
```

The expected behaviour would be that the counter for all rows is increased if you click on one of the two command links (since preserveRowState is unfortunately not available in ui:repeat - [JAVASERVERFACES_SPEC_PUBLIC-1230](https://java.net/jira/browse/JAVASERVERFACES_SPEC_PUBLIC-1230 "VDLDoc for <ui:repeat> does not have concept of rowStatePreserved which <h:dataTable> does have.")).

If you click only on the "ajax" links, this works as expected. If you click on the "submit" links, it also works - but the counter is reset to 0 (i.e. the state is lost) any time you switch between ajax/non-ajax.

Even worse, the same thing happens if you add an Ajax update _outside_ the ui:repeat. In this case, if the counter is increased with the "submit" links the state is preserved. Otherwise it is lost.

I believe this is a regression caused by #3837. This issue added the following code to UIRepeat to determine whether the rows should be rendered:

```
private boolean requiresRowIteration(VisitContext ctx) {
        boolean shouldIterate = !ctx.getHints().contains(VisitHint.SKIP_ITERATION); 
        if (!shouldIterate) {
            FacesContext faces = ctx.getFacesContext();  
            String sourceId = faces.getExternalContext().getRequestParameterMap().get("javax.faces.source");  
            boolean containsSource = sourceId != null ? sourceId.startsWith(super.getClientId(faces) + getSeparatorChar(faces)): false;  
            return containsSource;
        } else {
            return shouldIterate;
        }
    }
```

Thus it will iterate the ui:repeat rows with SKIP_ITERATION enabled if javax.faces.source was inside the ui:repeat.

The attached test application can be deployed on Glassfish 3.1.2, where it works as expected, and on Wildfly 10.1.0 (Mojarra 2.2.13) or Glassfish 5.0.b05 (Mojarra 2.3.0.m11), where the issue exists.

The component's save/restoreState log the client ID. When clicking the AJAX link inside the ui:repeat, this ID includes the row index:

```
form:j_idt10:0:inRepeat
```

In the other cases, when rows are not iterated, the row ID is not contained and thus state previously saved is lost:

```
form:j_idt10:inRepeat
```

The sample application also contains a dataTable with rowStatePreserved=false, that exhibits the behaviour I would expect of ui:repeat, and one with rowStatePreserved=true that has fully functional per-row state saving.

I realize that this is an issue in the "very fine print" of Mojarra, but it's causing us actual issues including data loss since we migrated to Mojarra 2.2.13.
#### Affected Versions
[2.2.13, 2.3.0-m10]
